lib/virtio/virtqueue.c

```diff
@@ -28,6 +29,69 @@ static int virtqueue_nused(struct virtqueue *vq);
 static int virtqueue_navail(struct virtqueue *vq);
 #endif
 
+#ifdef VIRTIO_D_CACHE
+enum vring_cache_op {
+	VRING_CACHE_FLUSH = 0,
+	VRING_CACHE_INVALIDATE = 1,
+};
+
+static inline void vring_control_block_flush(struct virtqueue *vq)
+{
+	struct vring *vr;
+	unsigned int size;
+
+	vr = &vq->vq_ring;
+	size = vring_size(vr->num, vr->align);
+
+	metal_cache_flush(vr->desc, size);
+}
+
+static inline void vring_desc_cache_op(struct virtqueue *vq,
```

Comment:
following functions look to me no very efficient, what about creating Macro to save the if condition?
 
#define VRING_NUM(virtqueue)  (vq->vq_ring->num)
#define VRING_ALIGN(virtqueue)  (vq->vq_ring->align)
#define VRING_DESC(virtqueue)  (vq->vq_ring->desc)

then instead of calling vring_desc_cache_op(vq,jVRING_CACHE_FLUSH)
just call  
metal_cache_flush(VRING_DESC(vq), vring_desc_size(VRING_NUM(vq)));

Reply:
ok